File: server/routes/transcribe.ts
import { RequestHandler } from "express";
import multer from "multer";
import FormData from "form-data";

// Configure multer for memory storage
const upload = multer({
  storage: multer.memoryStorage(),
  limits: {
    fileSize: 25 * 1024 * 1024, // 25MB limit (Whisper's max)
  },
});

// Middleware to handle file upload
export const uploadAudio = upload.single('audio');

// Handler for transcription
export const handleTranscribe: RequestHandler = async (req, res) => {
  try {
    // Check if file was uploaded
    if (!req.file) {
      return res.status(400).json({
        success: false,
        error: 'No audio file provided',
      });
    }

    // Create form data to send to AI Brain
    const formData = new FormData();
    formData.append('audio', req.file.buffer, {
      filename: req.file.originalname || 'recording.webm',
      contentType: req.file.mimetype,
    });
    formData.append('prompt', 'Transcribe this audio');

    // Forward to AI Brain service
    const aiBrainUrl = process.env.AI_BRAIN_URL || 'http://localhost:8001';
    const response = await fetch(`${aiBrainUrl}/router`, {
      method: 'POST',
      body: formData as any,
      headers: formData.getHeaders(),
    });

    if (!response.ok) {
      const errorText = await response.text();
      console.error('AI Brain transcription error:', errorText);
      
      return res.status(response.status).json({
        success: false,
        error: `Transcription service error: ${response.statusText}`,
      });
    }

    const data = await response.json();

    // Return transcribed text
    return res.json({
      success: true,
      text: data.response || data.text || '',
      model: data.model,
    });
  } catch (error) {
    console.error('Transcription error:', error);
    
    // Handle specific error types
    if (error instanceof Error) {
      if (error.message.includes('ECONNREFUSED')) {
        return res.status(503).json({
          success: false,
          error: 'AI Brain service is unavailable. Please ensure it is running.',
        });
      }
      
      if (error.message.includes('timeout')) {
        return res.status(504).json({
          success: false,
          error: 'Transcription request timed out. Please try again.',
        });
      }
    }

    return res.status(500).json({
      success: false,
      error: 'Internal server error during transcription',
    });
  }
};

File: python-backend/routers/transcribe.py
"""
Transcription router for handling audio transcription requests.
Forwards audio files to the AI Brain service for Whisper-based transcription.
"""

from fastapi import APIRouter, Request, HTTPException, UploadFile, File, Form
from fastapi.responses import JSONResponse
import logging
import httpx
from typing import Optional

from middleware.rate_limiter import limiter
from config import config

logger = logging.getLogger(__name__)

# Create router
router = APIRouter(prefix="/api", tags=["transcription"])


@router.post("/transcribe")
@limiter.limit(f"{config.RATE_LIMIT_REQUESTS}/{config.RATE_LIMIT_WINDOW}seconds")
async def transcribe_audio(
    request: Request,
    audio: UploadFile = File(...)
):
    """
    Transcribe audio file using AI Brain's Whisper integration.
    
    Accepts an audio file, forwards it to the AI Brain service at /router endpoint,
    and returns the transcribed text.
    
    Args:
        request: FastAPI request object (needed for rate limiting)
        audio: Audio file to transcribe (max 25MB as per Whisper requirements)
        
    Returns:
        JSON response with transcribed text or error message
        
    Raises:
        HTTPException: For various error conditions (400, 500, 503)
        
    Requirements: 3.1, 5.1, 5.2
    """
    try:
        # Validate file was uploaded
        if not audio:
            logger.warning("No audio file provided in transcription request")
            return JSONResponse(
                status_code=400,
                content={
                    "text": "",
                    "success": False,
                    "error": "No audio file provided"
                }
            )
        
        # Check file size (25MB limit for Whisper)
        file_content = await audio.read()
        file_size_mb = len(file_content) / (1024 * 1024)
        
        if file_size_mb > 25:
            logger.warning(f"Audio file too large: {file_size_mb:.2f}MB")
            return JSONResponse(
                status_code=400,
                content={
                    "text": "",
                    "success": False,
                    "error": f"Audio file too large ({file_size_mb:.2f}MB). Maximum size is 25MB."
                }
            )
        
        logger.info(
            f"Transcription request received - "
            f"Filename: {audio.filename}, "
            f"Content-Type: {audio.content_type}, "
            f"Size: {file_size_mb:.2f}MB"
        )
        
        # Prepare multipart form data for AI Brain
        files = {
            "audio": (audio.filename, file_content, audio.content_type)
        }
        data = {
            "prompt": "Transcribe this audio"
        }
        
        # Forward to AI Brain service
        ai_brain_url = f"{config.AI_BRAIN_ENDPOINT}/router"
        
        async with httpx.AsyncClient(timeout=120.0) as client:
            try:
                response = await client.post(
                    ai_brain_url,
                    files=files,
                    data=data
                )
                
                if response.status_code != 200:
                    logger.error(f"AI Brain service returned status {response.status_code}")
                    return JSONResponse(
                        status_code=503,
                        content={
                            "text": "",
                            "success": False,
                            "error": f"AI Brain service returned status {response.status_code}"
                        }
                    )
                
                # Parse AI Brain response
                response_data = response.json()
                transcribed_text = response_data.get("response", "") or response_data.get("text", "")
                
                logger.info(f"Transcription completed successfully - Text length: {len(transcribed_text)}")
                
                return JSONResponse(
                    status_code=200,
                    content={
                        "text": transcribed_text,
                        "success": True
                    }
                )
                
            except httpx.ConnectError as e:
                logger.error(f"Failed to connect to AI Brain service: {str(e)}")
                return JSONResponse(
                    status_code=503,
                    content={
                        "text": "",
                        "success": False,
                        "error": "AI Brain service is unavailable"
                    }
                )
                
            except httpx.TimeoutException as e:
                logger.error(f"Transcription request timed out: {str(e)}")
                return JSONResponse(
                    status_code=503,
                    content={
                        "text": "",
                        "success": False,
                        "error": "Transcription request timed out"
                    }
                )
                
    except Exception as e:
        # Handle unexpected errors
        logger.error(f"Unexpected error in transcription endpoint: {type(e).__name__}: {str(e)}", exc_info=True)
        return JSONResponse(
            status_code=500,
            content={
                "text": "",
                "success": False,
                "error": "Transcription failed"
            }
        )
